Wrappers/Python/cil/optimisation/functions/TotalGeneralisedVariation.py

```diff
@@ -0,0 +1,211 @@
+        >>> alpha0 = 2.0
+        >>> TGV = TotalGeneralisedVariation(alpha1 = alpha1, alpha0 = alpha0)
+        >>> sol = TGV.proximal(b, tau = 1.0)
+
+
+        """   
+        
+            
+    def __init__(self,
+                 reg_parameter=1.0,
+                 alpha1 = 1.0,
+                 alpha0 = 2.0,
+                 max_iteration = 100, 
+                 update_objective_interval = 50,
+                 correlation = "Space",
+                 backend = "c",
+                 split = False,
+                 verbose = 0,                  
+                 warmstart=False):
+        
+        super(TotalGeneralisedVariation, self).__init__(L = None)
+
+        # regularisation parameters for TGV
+        self.alpha1 = alpha1
+        self.alpha0 = alpha0
+        self.reg_parameter = reg_parameter
+                
+        # Iterations for PDHG_TGV
+        self.iterations = max_iteration
+        
+        # correlation space or spacechannels
+        self.correlation = correlation
+
+        # backend for the gradient
+        self.backend = backend        
+        
+        # splitting Gradient
+        self.split = split
+                        
+        self.verbose = verbose
+        self.update_objective_interval = update_objective_interval
+
+        # warm-start
+        self.warmstart  = warmstart
+        if self.warmstart:
+            self.hasstarted = False        
+
+
+    def __call__(self, x):
+        
+        if not hasattr(self, 'pdhg'):   
+            return 0.0        
+        else:              
+            # Compute alpha1 * || Du - w || + alpha0 * ||Ew||, 
+            # where (u,w) are solutions coming from the proximal method below.
+            tmp = self.f(self.pdhg.operator.direct(self.pdhg.solution))
+            return tmp
+
+    def _setup_solver(self, x, tau=1.0):
+
+        # tau_sqrt = np.sqrt(tau)
+        
+        if not hasattr(self, 'domain'):
+            
+            # sirf compatibility
+            try:
+                self.domain = x.geometry
+            except:
+                self.domain = x            
+                        
+        if not hasattr(self, 'operator'):
+            
+            self.Gradient = GradientOperator(self.domain, correlation = self.correlation, backend = self.backend)  
+            self.SymGradient = SymmetrisedGradientOperator(self.Gradient.range, correlation = self.correlation, backend = self.backend)  
+            self.ZeroOperator = ZeroOperator(self.domain, self.SymGradient.range)
+            self.IdentityOperator = - IdentityOperator(self.Gradient.range)
+
+            #    BlockOperator = [ Gradient      - Identity  ]
+            #                    [ ZeroOperator   SymGradient] 
+            self.operator = BlockOperator(self.Gradient, self.IdentityOperator, 
+                                          self.ZeroOperator,self.SymGradient,shape=(2,2)) 
+
+        if not all(hasattr(self, attr) for attr in ["g"]):
```

Comment:
With this caching of `g`, any future calls will ignore input `x` and return the proximal map of `self.g1.b`
Similarly with the caching of `f1` and `f2`, the input step length `tau` is ignored.

If caching is desired, I suggest adding an `else` statement to modify `self.g1.b`, `self.f1.scalar` and `self.f2.scalar` on future calls of `.proximal()`